### PR TITLE
Move the GC checksum from labels to annotations

### PR DESCRIPTION
**Pull request by JaneLiuL**
This PR moves the `kustomize.toolkit.fluxcd.io/checksum` from labels to annotations. This should fix any conflicts with 3rd party controllers like Stash (fix: #315) that are copying the labels from their custom resources to generated objects.

### CODE REVIEW

File: controllers/utils.go [hunk 1]
```diff
@@ -48,6 +48,53 @@ func parseApplyOutput(in []byte) map[string]string {
...
+// diff -u -N /tmp/LIVE-738467670/xx.v1alpha1.Keycloak.namespace.name /tmp/MERGED-747032253/xx.v1alpha1.Keycloak.namespace.name
+// --- /tmp/LIVE-738467670/xx.v1alpha1.Keycloak.namespace.name	2021-06-07 11:38:56.287331739 +0200
+// +++ /tmp/MERGED-747032253/xx.v1alpha1.Keycloak.namespace.name.test-mapping
+// @@ -8,7 +8,7 @@
+// ...
+// diff -u -N /tmp/LIVE-656588208/kustomize.toolkit.fluxcd.io.v1beta1.Kustomization.namespace.name /tmp/MERGED-532750671/kustomize.toolkit.fluxcd.io.v1beta1.Kustomization.namespace.name
+// --- /tmp/LIVE-656588208/kustomize.toolkit.fluxcd.io.v1beta1.Kustomization.namespace.name	2021-06-07 12:58:20.738794982 +0200
+// +++ /tmp/MERGED-532750671/kustomize.toolkit.fluxcd.io.v1beta1.Kustomization.namespace.name	2021-06-07 12:58:20.798795908 +0200
+// @@ -0,0 +1,36 @@
+func parasDiffOutput(in []byte) map[string]string {
```
Comment (stefanprodan) [inline controllers/utils.go:62]:
```suggestion
func parseDiffOutput(in []byte) map[string]string {
```

File: controllers/kustomization_gc.go [hunk 2]
```diff
@@ -138,7 +138,8 @@ func (kgc *KustomizeGarbageCollector) Prune(timeout time.Duration, name string,
 }
 
 func (kgc *KustomizeGarbageCollector) isStale(obj unstructured.Unstructured) bool {
-	itemChecksum := obj.GetLabels()[fmt.Sprintf("%s/checksum", kustomizev1.GroupVersion.Group)]
```
Comment (stefanprodan) [inline controllers/kustomization_gc.go:141]: We can't drop the label check in this version. If the label exists but the annotation doesn't, then we have to mark the object as stale otherwise after upgrading to this version the controller will fail to prune the objects reconciled with the older version. If both the label and annotation exists, then we should use the annotation.

**Comment by stefanprodan** [comment]
@JaneLiuL I found some serious issues with `diff`, for example, if a user adds namespaces and other objects in those namespaces, then the diff fails while the apply works. I see no way around this limitation but to detect diff failures, ignore them and switch back to the old way of detecting changes.

File: controllers/utils.go [hunk 3]
```diff
@@ -48,6 +48,50 @@ func parseApplyOutput(in []byte) map[string]string {
...
+// diff -u -N /tmp/LIVE-738467670/xx.v1alpha1.Keycloak.namespace.name /tmp/MERGED-747032253/xx.v1alpha1.Keycloak.namespace.name
+// --- /tmp/LIVE-738467670/xx.v1alpha1.Keycloak.namespace.name	2021-06-07 11:38:56.287331739 +0200
+// +++ /tmp/MERGED-747032253/xx.v1alpha1.Keycloak.namespace.name.test-mapping
+// @@ -8,7 +8,7 @@
+// ...
+// diff -u -N /tmp/LIVE-656588208/kustomize.toolkit.fluxcd.io.v1beta1.Kustomization.namespace.name /tmp/MERGED-532750671/kustomize.toolkit.fluxcd.io.v1beta1.Kustomization.namespace.name
+// --- /tmp/LIVE-656588208/kustomize.toolkit.fluxcd.io.v1beta1.Kustomization.namespace.name	2021-06-07 12:58:20.738794982 +0200
+// +++ /tmp/MERGED-532750671/kustomize.toolkit.fluxcd.io.v1beta1.Kustomization.namespace.name	2021-06-07 12:58:20.798795908 +0200
+// @@ -0,0 +1,36 @@
+func parseDiffOutput(in []byte) map[string]string {
```
Comment (stefanprodan) [inline controllers/utils.go:62]: Please remove this as it's unrelated to this PR.

File: controllers/kustomization_generator.go [hunk 4]
```diff
@@ -21,12 +21,10 @@ import (
 	"crypto/sha1"
 	"encoding/json"
 	"fmt"
+	"github.com/fluxcd/pkg/apis/kustomize"
```
Comment (stefanprodan) [inline controllers/kustomization_generator.go:24]: This needs to be added further down, the first import section is for Go std lib.

File: controllers/kustomization_generator.go [hunk 5]
```diff
@@ -86,18 +88,19 @@ func (kg *KustomizeGenerator) WriteFile(ctx context.Context, dirPath string) (st
...
 	} else {
-		var exists bool
-		for _, transformer := range kus.Transformers {
-			if transformer == transformerFileName {
-				exists = true
-				break
-			}
-		}
-		if !exists {
+		//kus.Transformers = append(kus.Transformers, transformerFileName)
```
Comment (stefanprodan) [inline controllers/kustomization_generator.go:93]: Please remove these commented lines.

File: controllers/kustomization_controller.go [hunk 6]
```diff
@@ -648,6 +649,9 @@ func (r *KustomizationReconciler) apply(ctx context.Context, kustomization kusto
 		}
 	}
 
+	changeSet := ""
```
Comment (stefanprodan) [inline controllers/kustomization_controller.go:652]: This change is unrelated to this PR please undo it.

File: controllers/kustomization_gc.go [hunk 7]
```diff
@@ -137,9 +137,25 @@ func (kgc *KustomizeGarbageCollector) Prune(timeout time.Duration, name string,
 	return changeSet, true
 }
 
+// We can't drop the label check in this version.
```
Comment (stefanprodan) [inline controllers/kustomization_gc.go:140]: Please replace this with:

> Check both labels and annotations for the checksum to preserve backwards compatibility

**Review by stefanprodan — COMMENTED** [review]
Please update the GC docs here https://github.com/fluxcd/kustomize-controller/blob/main/docs/spec/v1beta1/kustomization.md#garbage-collection

File: controllers/kustomization_generator.go [hunk 8]
```diff
@@ -277,13 +278,49 @@ func (kg *KustomizeGenerator) checksum(ctx context.Context, dirPath string) (str
...
+		return err
+	}
+
+	return nil
+}
+
 func (kg *KustomizeGenerator) generateLabelTransformer(checksum, dirPath string) error {
 	labels := selectorLabels(kg.kustomization.GetName(), kg.kustomization.GetNamespace())
 
-	// add checksum label only if GC is enabled
```
Comment (stefanprodan) [inline controllers/kustomization_generator.go:283]: This has been lost, please restore it for the annotions

File: controllers/kustomization_generator.go [hunk 9]
```diff
@@ -277,13 +278,49 @@ func (kg *KustomizeGenerator) checksum(ctx context.Context, dirPath string) (str
...
+			{Path: "metadata/annotations", CreateIfNotPresent: true},
+		},
+	}
+
+	data, err := yaml.Marshal(lt)
+	if err != nil {
+		return err
+	}
+
+	labelsFile := filepath.Join(dirPath, transformerAnnotationFileName)
```
Comment (stefanprodan) [inline controllers/kustomization_generator.go:312]:
```suggestion
	annotationsFile := filepath.Join(dirPath, transformerAnnotationFileName)
```

File: controllers/kustomization_generator.go [hunk 10]
```diff
@@ -36,13 +36,13 @@ import (
...
-	"github.com/fluxcd/pkg/apis/kustomize"
-
 	kustomizev1 "github.com/fluxcd/kustomize-controller/api/v1beta1"
+	"github.com/fluxcd/pkg/apis/kustomize"
 )
 
 const (
-	transformerFileName = "kustomization-gc-labels.yaml"
+	transformerFileName           = "kustomization-gc-labels.yaml"
+	transformerAnnotationFileName = "kustomization-annotation-labels.yaml"
```
Comment (stefanprodan) [inline controllers/kustomization_generator.go:45]:
```suggestion
	transformerAnnotationFileName = "kustomization-gc-annotations.yaml"
```

Comment (JaneLiuL) [reply controllers/kustomization_generator.go]: Fix done~

File: controllers/kustomization_generator.go (hunk 9 above)
Comment (JaneLiuL) [reply controllers/kustomization_generator.go]: Fix done~

File: controllers/kustomization_gc.go (hunk 7 above)
Comment (JaneLiuL) [reply controllers/kustomization_gc.go]: Fix done~

**Review by stefanprodan — APPROVED** [review]
LGTM

Thanks @JaneLiuL 🌷 

PS. I'll merge this and do a followup PR to update the GC documentation.
